homeassistant/components/mobile_app/entity.py

```diff
@@ -0,0 +1,135 @@
+"""A entity class for mobile_app."""
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorDevice
+from homeassistant.const import (CONF_WEBHOOK_ID, STATE_ON, STATE_OFF)
+from homeassistant.core import callback
+from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers.entity import Entity
+
+from .const import (ATTR_SENSOR_ATTRIBUTES,
+                    ATTR_SENSOR_DEVICE_CLASS, ATTR_SENSOR_ICON,
+                    ATTR_SENSOR_NAME, ATTR_SENSOR_STATE, ATTR_SENSOR_TYPE,
+                    ATTR_SENSOR_TYPE_BINARY_SENSOR, ATTR_SENSOR_UNIQUE_ID,
+                    ATTR_SENSOR_UOM, DATA_REGISTRATIONS, DOMAIN,
+                    SIGNAL_SENSOR_UPDATE)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_mobile_app_platform(sensor_type, hass, config,
+                                          async_add_entities,
+                                          discovery_info=None):
+    """Set up the mobile app entity sensor."""
+    registrations = hass.data[DOMAIN][DATA_REGISTRATIONS]
+    sensor_configs = hass.data[DOMAIN][sensor_type]
+
+    platform_name = "MobileAppEntity"
+
+    if sensor_type == ATTR_SENSOR_TYPE_BINARY_SENSOR:
+        platform_name = "MobileAppBinarySensor"
+
+    platform = globals()[platform_name]
+
+    entities = list()
+
+    if discovery_info is None:
+        for sensor_config in sensor_configs.values():
+            device = registrations[sensor_config[CONF_WEBHOOK_ID]]
+            entities.append(platform(sensor_config, device))
+    else:
+        device = registrations[discovery_info[CONF_WEBHOOK_ID]]
+        sensor_config = sensor_configs[discovery_info[ATTR_SENSOR_UNIQUE_ID]]
+        entities.append(platform(sensor_config, device))
+
+    async_add_entities(entities, True)
+
+
+class MobileAppEntity(Entity):
+    """Representation of an mobile app entity."""
+
+    def __init__(self, config, device):
+        """Initialize the sensor."""
+        self._config = config
+        self._device = device
+        self._device_id = device[CONF_WEBHOOK_ID]
+        self._sensor_id = "{}_{}".format(device[CONF_WEBHOOK_ID],
+                                         config[ATTR_SENSOR_UNIQUE_ID])
+        self._entity_type = config[ATTR_SENSOR_TYPE]
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        async_dispatcher_connect(self.hass, SIGNAL_SENSOR_UPDATE,
+                                 self._handle_update)
+
+    async def async_will_remove_from_hass(self):
+        """Disconnect dispatcher listener when removed."""
+        async_dispatcher_connect(self.hass, SIGNAL_SENSOR_UPDATE,
```

Comment:
Shouldn't we unregister the signal here?

Reply:
It's annoyingly confusing, but `async_dispatcher_connect` will both register/connect a signal as well as deregister/disconnect a signal if called again.

Reply:
`async_dispatcher_connect` returns a function that when called deregisters the signal. We're not calling that function here.

Reply:
Good catch!